nemo/collections/asr/data/text_to_text.py

```diff
@@ -0,0 +1,462 @@
+
+
+class TextToTextItem(NamedTuple):
+    tts_text: torch.Tensor  # normalized and tokenized text for TTS
+    transcript: torch.Tensor  # tokenized text for ASR
+    speaker: int  # speaker id for multi-speaker TTS
+
+
+class TextToTextBatch(NamedTuple):
+    tts_texts: torch.Tensor  # tokenized texts for tts
+    tts_text_length: torch.Tensor
+    transcripts: torch.Tensor  # tokenized texts for ASR
+    transcript_length: torch.Tensor
+    speakers: torch.Tensor  # speaker ids for multi-speaker TTS
+
+    @staticmethod
+    def collate_fn(batch: List[TextToTextItem], asr_pad_id: int, tts_text_pad_id: int) -> TextToTextBatch:
+        return TextToTextBatch(
+            tts_texts=pad_sequence([item.tts_text for item in batch], batch_first=True, padding_value=tts_text_pad_id),
+            tts_text_length=torch.tensor([item.tts_text.shape[0] for item in batch]).long(),
+            transcripts=pad_sequence([item.transcript for item in batch], batch_first=True, padding_value=asr_pad_id),
+            transcript_length=torch.tensor([item.transcript.shape[0] for item in batch]).long(),
+            speakers=torch.tensor([item.speaker for item in batch]).long(),
+        )
+
+
+class TextOrAudioToTextBatch(NamedTuple):
+    audio_signal: torch.Tensor
+    a_sig_length: torch.Tensor
+    tts_texts: torch.Tensor
+    tts_text_length: torch.Tensor
+    speakers: torch.Tensor
+    transcripts: torch.Tensor
+    transcript_length: torch.Tensor
+
+    @staticmethod
+    def collate_fn(
+        batch: List[Union[TextToTextItem, tuple]], tts_text_pad_id: int, asr_pad_id: int
+    ) -> Union[TextToTextBatch, TextOrAudioToTextBatch, tuple]:
+        """
+        Collate function for dataloader
+        Can accept mixed batch of text-to-text items and audio-text items (typical for ASR)
+        """
+        text_items: List[TextToTextItem] = [item for item in batch if isinstance(item, TextToTextItem)]
+        if not text_items:
+            # pure audio-text batch
+            return _speech_collate_fn(batch=batch, pad_id=asr_pad_id)
+
+        asr_items = [item for item in batch if not isinstance(item, TextToTextItem)]
+
+        if not asr_items:
+            # pure text-to-text batch
+            return TextToTextBatch.collate_fn(batch=text_items, asr_pad_id=asr_pad_id, tts_text_pad_id=tts_text_pad_id)
+
+        # mixed batch
+        audio_signal = pad_sequence([item[0] for item in asr_items], batch_first=True, padding_value=0.0)
+        a_sig_length = torch.tensor([item[1] for item in asr_items]).long()
+
+        tts_texts = pad_sequence(
+            [item.tts_text for item in text_items], batch_first=True, padding_value=tts_text_pad_id
+        )
+        tts_text_length = torch.tensor([item.tts_text.shape[0] for item in text_items]).long()
+        speakers = torch.tensor([item.speaker for item in text_items]).long()
+
+        transcripts = pad_sequence(
+            [item.transcript for item in text_items] + [item[2] for item in asr_items],
```

Comment:
Is indexing like `item[2]` necessary?
Can this be made more self-explanatory?
Or add some comments

Reply:
Added a comment about ASR items. Since they are tuples, to use generators (faster than looping) we have to use indices. It would be great to rewrite ASR items/batches with named tuples, but this is out of scope of this PR.